transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java

```diff
@@ -275,6 +284,28 @@ public boolean isActive() {
         return inboundMessages;
     }
 
+    public Queue<Object> tempInboundMessages() {
+        if (tempInboundMessages == null) {
+            tempInboundMessages = new ConcurrentLinkedQueue<Object>();
+        }
+        return tempInboundMessages;
+    }
+
+    private void drainTempInboundMessages() {
```

Comment:
Since both `writeOneInbound` and `channel.read()` can drain `tempInboundMessages`, I added a lock to ensure that only one thread can drain the queue to ensure order.

Reply:
`EmbeddedChannel` is not meant to be used in a multi-threaded way. It is expected to be called from a single thread only